src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,147 @@
+            tail = tail.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        indexChecker(index);
+        final Node<T> temporary = new Node<>(null, value, null);
+        if (index == 0) {
+            temporary.next = head;
+            head.prev = temporary;
+            head = temporary;
+            size++;
+            return;
+        }
+        Node<T> current = getByIndex(index);
+        temporary.prev = current.prev;
+        temporary.next = current;
+        current.prev.next = temporary;
+        current.prev = temporary;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> element = getByIndex(index);
+        T oldElement = element.value;
+        element.value = value;
+        return oldElement;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> elementToRemove = getByIndex(index);
+        removeLink(elementToRemove);
+        return elementToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (object == current.value || (object != null && object.equals(current.value))) {
+                removeLink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public Node<T> getByIndex(int index) {
+        indexChecker(index);
+        Node<T> current;
+        if (size / 2 > index) {
+            current = head;
+            for (int i = 0;i < index;i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size; i > (index + 1);i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private T removeLink(Node<T> node) {
+        if (node == head) {
+            if (size != 1) {
+                head.next.prev = null;
+                head = head.next;
+            } else {
+                head = null;
+                tail = null;
+            }
+        } else if (node == tail) {
+            tail.prev.next = null;
+            tail = tail.prev;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
```

Comment:
maybe will be more simple to write like this:
```suggestion
        if (node == head) {
            node.prev.next = node.next;
        } else {
            head = node.next;
        }
        if (node.next== tail) {
            node.next.prev = node.prev;
        } else {
            tail = node.prev;
        }
        size--;
```

Reply:
@AnonimToster good suggestion.